numpy/lib/function_base.py

```diff
@@ -4305,80 +4524,148 @@ def _percentile(a, q, axis=None, out=None,
+                                    prior_cum_w[nonzero_w_inds]) /
+                                   ws_sorted[nonzero_w_inds])
+
+        w_slice = np.roll(w_slice, -1, axis=-1)
+        # now create the lower percentage bound
+        normalized_w_lower = np.roll(normalized_w_upper + w_slice, 1, axis=-1)
+        normalized_w_lower[..., 0] = 0.0
+
+        # now we substract by left-most w_slice value
+        new_w_upper = (normalized_w_upper.T - w_slice[..., 0].T).T
+        new_w_upper[new_w_upper < 0.0] = 0.0
+        new_w_lower = (normalized_w_lower.T - w_slice[..., 0].T).T
+        new_w_lower[new_w_lower < 0.0] = 0.0
+        new_w_lower[..., 0] = 0.0
+
+        # renormalize by right-most bound
+        normalized_w_upper = (new_w_upper.T / new_w_upper[..., -1].T).T
+        normalized_w_lower = (new_w_lower.T / new_w_upper[..., -1].T).T
+
+        # combine and resort
+        cum_w_bands = np.concatenate([normalized_w_upper, normalized_w_lower],
+                                     axis=-1)
+        inds_resort = np.argsort(cum_w_bands, axis=-1)
+        cum_w_bands = arraysort(cum_w_bands, inds_resort)
+
+        ap = np.concatenate([ap_sorted, ap_sorted], axis=-1)
+        ap = arraysort(ap, inds_resort)
+
+        # interpolate
+        Nx = ap.shape[-1]
+        indices_hard = np.arange(Nx)
+        vec_interp_func = np.vectorize(np.interp, signature='(n),(m),(m)->(n)')
+        indices = vec_interp_func(q, cum_w_bands, indices_hard)
 
-    # round fractional indices according to interpolation method
     if interpolation == 'lower':
-        indices = floor(indices).astype(intp)
+        indices = np.floor(indices).astype(intp)
     elif interpolation == 'higher':
-        indices = ceil(indices).astype(intp)
+        indices = np.ceil(indices).astype(intp)
     elif interpolation == 'midpoint':
-        indices = 0.5 * (floor(indices) + ceil(indices))
+        indices = 0.5 * (np.floor(indices) + np.ceil(indices))
     elif interpolation == 'nearest':
-        indices = around(indices).astype(intp)
+        indices = np.around(indices).astype(intp)
     elif interpolation == 'linear':
         pass  # keep index as fraction and interpolate
     else:
         raise ValueError(
             "interpolation can only be 'linear', 'lower' 'higher', "
             "'midpoint', or 'nearest'")
 
-    n = np.array(False, dtype=bool) # check for nan's flag
-    if indices.dtype == intp:  # take the points along axis
-        # Check if the array contains any nan's
-        if np.issubdtype(a.dtype, np.inexact):
-            indices = concatenate((indices, [-1]))
+    if indices.dtype == intp and weights is None:  # take the points along axis
 
-        ap.partition(indices, axis=axis)
-        # ensure axis with qth is first
-        ap = np.rollaxis(ap, axis, 0)
-        axis = 0
+        def _take1d(vec, inds):
+            return take(vec, inds)
+
+        vec_take = np.vectorize(_take1d, signature='(n),(m)->(m)')
```

Comment:
That might work, but I think having separate names (and `signature`) provides more clarity on what these two routines are trying to accomplish.  I think It certainly helps understanding/debugging the code.  Unless there's a real performance impact I prefer to keep it the way it is, though I'd defer to your judgment.